libevm/VMSIMD.cpp

```diff
@@ -0,0 +1,687 @@
+#define ROR( x1, x2) (((x1) >> x2)|((x1) << (sizeof(x1) * 8 - (x2))))
+
+void VM::xadd (uint8_t _b) { EVALXOPU(ADD, _b); }
+void VM::xmul (uint8_t _b) { EVALXOPU(MUL, _b); }
+void VM::xsub (uint8_t _b) { EVALXOPU(SUB, _b); }
+void VM::xdiv (uint8_t _b) { EVALXOPU(DIV, _b); }
+void VM::xsdiv(uint8_t _b) { EVALXOPS(DIV, _b); }
+void VM::xmod (uint8_t _b) { EVALXOPU(MOD, _b); }
+void VM::xsmod(uint8_t _b) { EVALXOPS(MOD, _b); }
+void VM::xlt  (uint8_t _b) { EVALXOPU(LT,  _b); }
+void VM::xslt (uint8_t _b) { EVALXOPS(LT,  _b); }
+void VM::xgt  (uint8_t _b) { EVALXOPU(GT,  _b); }
+void VM::xsgt (uint8_t _b) { EVALXOPS(GT,  _b); }
+void VM::xeq  (uint8_t _b) { EVALXOPU(EQ,  _b); }
+void VM::xzero(uint8_t _b) { EVALXOPU(ZERO,_b); }
+void VM::xand (uint8_t _b) { EVALXOPU(AND, _b); }
+void VM::xoor (uint8_t _b) { EVALXOPU(OR,  _b); }
+void VM::xxor (uint8_t _b) { EVALXOPU(XOR, _b); }
+void VM::xnot (uint8_t _b) { EVALXOPU(NOT, _b); }
+void VM::xshr (uint8_t _b) { EVALXOPU(SHR, _b); }
+void VM::xsar (uint8_t _b) { EVALXOPS(SHR, _b); }
+void VM::xshl (uint8_t _b) { EVALXOPU(SHL, _b); }
+void VM::xrol (uint8_t _b) { EVALXOPU(ROL, _b); }
+void VM::xror (uint8_t _b) { EVALXOPU(ROR, _b); }
+
+inline uint8_t pow2N(uint8_t n)
+{
+	static uint8_t exp[6] = { 1, 2, 4, 8, 16, 32 };
+	return exp[n];
+}
+
+inline uint8_t nElem(uint8_t _b)
+{
+	return pow2N((_b) & 0xf);
+}
+
+inline uint8_t elemType(uint8_t _b)
+{
+	return (_b) >> 4;
+
+
+void VM::vtow(uint8_t _b, const u256& _in, u256& _out)
+{
+	const uint8_t n = nElem(_b), t = elemType(_b);
+	switch (t)
+	{
+	case 0:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			_out << 8;
+			_out |= v8x32(_in) [i];
+		}
+		break;
+	case 1:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			_out << 16;
+			_out |= v16x16(_in)[i];
+		}
+		break;
+	case 2:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			_out << 32;
+			_out |= v32x8(_in) [i];
+		}
+		break;
+	case 3:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			_out << 64;
+			_out |= v64x4(_in) [i];
+		}
+		break;
+	default: throwBadInstruction();
+	}
+}
+
+void VM::wtov(uint8_t _b, const u256& _in, u256& _out)
+
+{
+	const uint8_t n = nElem(_b), t = elemType(_b);
+	switch (t)
+	{
+	case 0:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			v8x32(_out) [i] |= (uint8_t )(_in & 0xff);
+			_in  << 8;
+		}
+		break;
+	case 1:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			v16x16(_out)[i] |= (uint16_t)(_in & 0xffff);
+			_in << 16;
+		}
+		break;
+	case 2:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			v32x8(_out) [i] |= (uint32_t)(_in & 0xffffff);
+			_in << 32;
+		}
+		break;
+	case 3:
+		for (int i = n-1; 0 <= i; --i)
+		{ 
+			v64x4(_out) [i] |= (uint64_t)(_in & 0xffffffff);
+			_in << 64;
+		}
+		break;
+	default: throwBadInstruction();
+	}
+}
+
+void VM::xmload (uint8_t _b)
+{
+	// n bytes of type t elements in memory vector
+	// goes onto stack element by element, LSB first
+	uint8_t *p = m_mem.data() + toInt15(m_SP[0]);
+	const uint8_t n = nElem(_b), t = elemType(_b);
```

Comment:
I think that only works on LSB-first machines, and there is no portable way to know whether you are on one.  So that becomes part of the non-portable optimizations.